### PR TITLE
Update data generator to support folders, assay designs, and storage hierarchy

### PR DESCRIPTION
**Pull request by labkey-susanh**
#### Rationale
Here we beef up our capabilities for generating data at scale to include the generation of folders, assay designs, and storage hierarchies. This PR does not include the generation of data within any of these entities, only the definition of the entities themselves.

#### Related Pull Requests
- https://github.com/LabKey/inventory/pull/707
- https://github.com/LabKey/biologics/pull/1891

#### Changes
* add `DataGeneratorRegistry` and `DataGenerationDriver` interface
* add `generateFolders` method to `DataGenerator`
* add `AssayDesignGenerator`
* In `createContainerList`, don't add the current container more than once.

### CODE REVIEW

File: api/src/org/labkey/api/data/generator/DataGenerator.java [hunk 1]
```diff
@@ -174,14 +178,46 @@ public void setCustomDataClasses(List<ExpDataClass> customDataClasses)
...
+        checkAlive(_job);
+        int numFolders = _config.getNumFolders();
+        if (numFolders <= 0)
+        {
+            _log.info(String.format("No folders generated because %s=%d", Config.NUM_FOLDERS, numFolders));
+            return;
+        }
+        CPUTimer timer = addTimer(String.format("%d sub-folders", numFolders));
+        timer.start();
+        Set<String> currentChildren = ContainerManager.getChildren(getContainer()).stream().map(Container::getName).collect(Collectors.toSet());
```
Comment (labkey-nicka) [inline api/src/org/labkey/api/data/generator/DataGenerator.java]: Should this be a case-insensitive Set?

File: api/src/org/labkey/api/data/generator/DataGeneratorRegistry.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Map;
+
+public class DataGeneratorRegistry
+{
+    public enum DataType {
+        AssayDesigns,
+        Storage,
+    };
+
+    private static Map<DataType, DataGenerator.DataGenerationDriver> _dataGeneratorMap = new HashMap<>();
```
Comment (labkey-nicka) [inline api/src/org/labkey/api/data/generator/DataGeneratorRegistry.java]: nit: Can be `final`.

File: api/src/org/labkey/api/data/generator/DataGenerator.java [hunk 3]
```diff
@@ -637,6 +682,7 @@ public static long randomLong(long startIndex, long endIndex)
         public static final String MIN_NUM_FIELDS = "minFields";
         public static final String MAX_NUM_FIELDS = "maxFields";
 
+        int _numFolders = 0;
```
Comment (labkey-nicka) [inline api/src/org/labkey/api/data/generator/DataGenerator.java]: Why do these have default values provided here (as opposed to the default values in the constructor) where as you've elected to remove the default values in the `BiologicsDataGenerator`?

**Review by labkey-nicka — APPROVED** [review]

File: api/src/org/labkey/api/data/generator/DataGeneratorRegistry.java (hunk 2 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/data/generator/DataGeneratorRegistry.java]: Thanks. Done.

File: api/src/org/labkey/api/data/generator/DataGenerator.java (hunk 3 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/data/generator/DataGenerator.java]: 'cause I forgot to update this class. Done.

File: api/src/org/labkey/api/data/generator/DataGenerator.java (hunk 1 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/data/generator/DataGenerator.java]: Yes, I guess we are case-insensitive when checking for folder name overlaps.  Done.

**Pull request merged**